### DISCUSSION TITLE
Azure AD OAuth & Refresh Tokens

### DISCUSSION THREAD

**Discussion by Jaiieq** [PHPMailer/PHPMailer#2927] [General]
I have a quick question to possibly clear up some confusion or misconceptions on my part.

First of all I have Azure AD OAuth and PHPmailer working great;  Sending mails no problem.  My question is regarding the Refresh Token.  Is there not a mechanism (or perhaps some config I have missed) that auto-updates the Refresh Token?  Currently, I'm having to do it manually every 90 days using the get_oauth_token script when it expires.

Any help or insight would be greatly appreciated!

**Comment by decomplexity** [comment]
Assuming you are using TheLeague’s OAuth2 code, you need a one-line code addition is needed to TheLeague oauth2-client/src/Token/AccessToken.php.
The existing code from around line 107 (exact line is version dependent) reads:

```
        if (!empty($options['refresh_token'])) {
            $this->refreshToken = $options['refresh_token'];
```

Then immediately after this line add the line:

```
        $_SESSION[__NAMESPACE__ . "\\updatedRefreshToken"] =  $this->refreshToken; 
```

that stores the token as a session variable and tries to ensure that the variable name is not used elsewhere.

Then retrieve it in your own global class or code with something like:

```
public function storeNewRefreshToken() 
    {
      /** 
	check if key exists
        */
        $key = "League\\OAuth2\\Client\\Token" . "\\updatedRefreshToken";
        if (array_key_exists($key,$_SESSION))
        {
        $nrt = $_SESSION[$key];

       /** 
         check if value was not empty
         */
        if (!empty($nrt)) {
	   
        /**
          * store the token for retrieval and use next time you invoke PHPMailer with your global code:        
          * write the token to file with e.g. a
          *    file_put_contents(. . . .);
          * or store in a database
	  */
          } // ends 'if token was found but value was not empty'
       } // ends 'if a token was actually found in  $_SESSION
    }
```

**Comment by Jaiieq** [comment]
Thank you.  I'll give this a go!

**Reply by fischel** [reply]
@Jaiieq  did it work for you? So you get a new refresh_token? Which provider are you using? thx for help

**Reply by br4nnigan** [reply]
> Thank you. I'll give this a go!

I would appreciate a confirmation too. I'm not really seeing the connection between the question and the answer. I'm using Greew\OAuth2\Client\Provider\Azure and of course I'm already storing the refresh token in the db to use it. Which works, for 90 days ...

**Comment by MattAllatt** [comment]
I know this discussion stopped a year ago, but I was looking at the same problem of capturing new refresh tokens, and came up with a slightly different solution. The $_SESSION solution doesn't work if you're running the PHP script from command line, or a cron job, etc.

I solved the problem by adding a simple getter to the bottom of PHPMailer's OAuth.php file:
```
    public function getOauthToken()
    {
        return $this->oauthToken;
    }
```

and then after the send, you can get to the new refresh token with the new getter:
```
        $mail->send();
        $new_refresh_token = $mail->getOAuth()->getOauthToken()->getRefreshToken();
        if($new_refresh_token !== $old_refresh_token && $new_refresh_token !== "")
        {
/* Update your records with the new refresh token */
        }

```
